File: hackerrank/challenges/Functional Programming/Recursion/haskell/sequence_full_of_colors.hs
import Control.Monad

fullOfColors [] r g y b = r==g && y==b
fullOfColors (x:xs) r g y b
  | abs (r-g) > 1 = False
  | abs (y-b) > 1 = False
  | otherwise = fullOfColors xs (r+rcount) (g+gcount) (y+ycount) (b+bcount)
  where rcount = if x == 'R' then 1 else 0
        gcount = if x == 'G' then 1 else 0
        ycount = if x == 'Y' then 1 else 0
        bcount = if x == 'B' then 1 else 0

main = do
  t <- readLn
  replicateM_ t $ do
     s <- getLine
     print $ fullOfColors s 0 0 0 0

-- main = do
--   print $ fullOfColors "RGGR" 0 0 0 0
--   print $ fullOfColors "RYBG" 0 0 0 0
--   print $ fullOfColors "RYRB" 0 0 0 0
--   print $ fullOfColors "YGYGRBRB" 0 0 0 0

File: hackerrank/challenges/Functional Programming/Recursion/python3/sequence_full_of_colors.py
#!python3.5

def full_of_colors(seq):
  r = 0
  g = 0
  y = 0
  b = 0

  for i in seq:
    if i == "R":
      r+=1
    elif i == "G":
      g+=1
    elif i == "Y":
      y+=1
    elif i == "B":
      b+=1

    if abs(r-g) > 1 or abs(y-b) > 1:
      return False

  return r-g == 0 and y-b == 0

print(full_of_colors("RGGR"))
print(full_of_colors("RYBG"))
print(full_of_colors("RYRB"))
print(full_of_colors("YGYGRBRB"))
